### ISSUE TITLE
CohortPackager error message when consuming ExtractedFileStatusMessage from CTP

### ISSUE THREAD

**Issue by howff** [SMI/SmiServices#942]
I've just run an extraction and noticed a handful of messages were in the ExtractFile**No**VerifyQueue which I thought was odd. Maybe a problem with CTP or the anonymiser?

After running CohortPackager the queues were all empty, except for a handful in the DLQueue. They are from CTPAnonymiserHost but just say Status:ErrorWontRetry, no other reason for failure.  I can check the logs but haven't done so yet.

Anyway CohortPackager appears to have finished (i.e. no more output to screen), but it has **not** produced any reports.

It has also dumped a lot of error messages:

```
$   ./smi cohort-packager -y /nfs/smi/home/smi/configs/smi_dataExtract.yaml
Bootstrapper -> Main called, constructing host
2021-10-28 20:22:34.5624| INFO|SmiLogging|Logging config loaded from /nfs/smi/home/smi/configs/Smi.NLog.config||
2021-10-28 20:22:34.5749| INFO|CohortPackagerHost|Host logger created||
2021-10-28 20:22:34.5762| INFO|CohortPackagerHost|Starting CohortPackager (Host=nsh-smi04 PID=7977 User=abrooks)||
2021-10-28 20:22:35.3485| INFO|MongoExtractJobStore|Connected to job store with 4 existing jobs||
2021-10-28 20:22:35.3549| WARN|FileReporter|Not passed a specific newline string for creating reports. Defaulting to Environment.NewLine ('\n')||
Bootstrapper -> Host constructed, starting aux connections
Bootstrapper -> Host aux connections started, calling Start()
Bootstrapper -> Host created and started...
MicroserviceHostBootstrapper exiting. Service will exit when consumer threads are joined
2021-10-28 20:22:51.8023| INFO|MongoExtractJobStore|Received new job info ExtractionJobIdentifier=29a5e4a7-8091-44c0-a726-e620dce3587e, ProjectNumber=1516-IMG3, ExtractionDirectory=1516-IMG3/extractions/SeriesInstanceUID_for_SR_MR_chest, JobSubmittedAt=2021-10-28T16:50:43, IsIdentifiableExtraction=False, IsNoFilterExtraction=False, ,KeyTag=SeriesInstanceUID,KeyValueCount=100,ExtractionModality=||
2021-10-28 20:22:51.8090|ERROR|AnonFailedMessageConsumer|Could not deserialize message to ExtractedFileStatusMessage object. Likely an issue with the message content|Newtonsoft.Json.JsonSerializationException: Couldn't deserialize message to Smi.Common.Messages.Extraction.ExtractedFileStatusMessage. See exception data.
   at Smi.Common.MessageSerialization.JsonConvert.DeserializeObject[T](String message) in /home/vsts/work/1/s/src/common/Smi.Common/MessageSerialization/JsonConvert.cs:line 52
   at Smi.Common.MessageSerialization.JsonConvert.DeserializeObject[T](BasicDeliverEventArgs deliverArgs) in /home/vsts/work/1/s/src/common/Smi.Common/MessageSerialization/JsonConvert.cs:line 69
   at Smi.Common.Messaging.Consumer`1.SafeDeserializeToMessage[T](IMessageHeader header, BasicDeliverEventArgs deliverArgs, T& iMessage) in /home/vsts/work/1/s/src/common/Smi.Common/Messaging/Consumer.cs:line 139 0: Required property 'StatusMessage' not found in JSON. Path '', line 1, position 434.|MessageGuid=f999a65b-0083-4421-b24a-5d2b2b13a20e, ProducerProcessID=1081514896, ProducerExecutableName=CTPAnonymiserHost, OriginalPublishTimestamp=1635439874, Parents=ed762b00-dbba-4239-b645-3d914b7531f6->7772dc72-3cc6-4559-a429-c55f7255f7f3
2021-10-28 20:22:51.8241|ERROR|AnonFailedMessageConsumer|Could not deserialize message to ExtractedFileStatusMessage object. Likely an issue with the message content|Newtonsoft.Json.JsonSerializationException: Couldn't deserialize message to Smi.Common.Messages.Extraction.ExtractedFileStatusMessage. See exception data.
   at Smi.Common.MessageSerialization.JsonConvert.DeserializeObject[T](String message) in /home/vsts/work/1/s/src/common/Smi.Common/MessageSerialization/JsonConvert.cs:line 52
   at Smi.Common.MessageSerialization.JsonConvert.DeserializeObject[T](BasicDeliverEventArgs deliverArgs) in /home/vsts/work/1/s/src/common/Smi.Common/MessageSerialization/JsonConvert.cs:line 69
   at Smi.Common.Messaging.Consumer`1.SafeDeserializeToMessage[T](IMessageHeader header, BasicDeliverEventArgs deliverArgs, T& iMessage) in /home/vsts/work/1/s/src/common/Smi.Common/Messaging/Consumer.cs:line 139 0: Required property 'StatusMessage' not found in JSON. Path '', line 1, position 434.|MessageGuid=7b580519-f6c9-48aa-ba10-47ba7656fc3e, ProducerProcessID=1081514896, ProducerExecutableName=CTPAnonymiserHost, OriginalPublishTimestamp=1635439874, Parents=ed762b00-dbba-4239-b645-3d914b7531f6->7dfbe249-2b92-481c-ac0c-f319ec977a08
2021-10-28 20:22:51.8248|ERROR|AnonFailedMessageConsumer|Could not deserialize message to ExtractedFileStatusMessage object. Likely an issue with the message content|Newtonsoft.Json.JsonSerializationException: Couldn't deserialize message to Smi.Common.Messages.Extraction.ExtractedFileStatusMessage. See exception data.
   at Smi.Common.MessageSerialization.JsonConvert.DeserializeObject[T](String message) in /home/vsts/work/1/s/src/common/Smi.Common/MessageSerialization/JsonConvert.cs:line 52
   at Smi.Common.MessageSerialization.JsonConvert.DeserializeObject[T](BasicDeliverEventArgs deliverArgs) in /home/vsts/work/1/s/src/common/Smi.Common/MessageSerialization/JsonConvert.cs:line 69
   at Smi.Common.Messaging.Consumer`1.SafeDeserializeToMessage[T](IMessageHeader header, BasicDeliverEventArgs deliverArgs, T& iMessage) in /home/vsts/work/1/s/src/common/Smi.Common/Messaging/Consumer.cs:line 139 0: Required property 'StatusMessage' not found in JSON. Path '', line 1, position 434.|MessageGuid=a9d188bd-0ecb-4f68-8b93-0ac991de2e58, ProducerProcessID=1081514896, ProducerExecutableName=CTPAnonymiserHost, OriginalPublishTimestamp=1635439874, Parents=ed762b00-dbba-4239-b645-3d914b7531f6->e8e21ba8-a081-4607-a2a6-e311c66a8e1d
2021-10-28 20:22:51.8248|ERROR|AnonFailedMessageConsumer|Could not deserialize message to ExtractedFileStatusMessage object. Likely an issue with the message content|Newtonsoft.Json.JsonSerializationException: Couldn't deserialize message to Smi.Common.Messages.Extraction.ExtractedFileStatusMessage. See exception data.
   at Smi.Common.MessageSerialization.JsonConvert.DeserializeObject[T](String message) in /home/vsts/work/1/s/src/common/Smi.Common/MessageSerialization/JsonConvert.cs:line 52
   at Smi.Common.MessageSerialization.JsonConvert.DeserializeObject[T](BasicDeliverEventArgs deliverArgs) in /home/vsts/work/1/s/src/common/Smi.Common/MessageSerialization/JsonConvert.cs:line 69
   at Smi.Common.Messaging.Consumer`1.SafeDeserializeToMessage[T](IMessageHeader header, BasicDeliverEventArgs deliverArgs, T& iMessage) in /home/vsts/work/1/s/src/common/Smi.Common/Messaging/Consumer.cs:line 139 0: Required property 'StatusMessage' not found in JSON. Path '', line 1, position 434.|MessageGuid=e23569d4-ecce-402f-aeb7-ccba4ab3528e, ProducerProcessID=1081514896, ProducerExecutableName=CTPAnonymiserHost, OriginalPublishTimestamp=1635439874, Parents=ed762b00-dbba-4239-b645-3d914b7531f6->b6dadbe5-996b-464f-b126-9f85e9638d2f
2021-10-28 20:22:51.8259|ERROR|AnonFailedMessageConsumer|Could not deserialize message to ExtractedFileStatusMessage object. Likely an issue with the message content|Newtonsoft.Json.JsonSerializationException: Couldn't deserialize message to Smi.Common.Messages.Extraction.ExtractedFileStatusMessage. See exception data.
   at Smi.Common.MessageSerialization.JsonConvert.DeserializeObject[T](String message) in /home/vsts/work/1/s/src/common/Smi.Common/MessageSerialization/JsonConvert.cs:line 52
   at Smi.Common.MessageSerialization.JsonConvert.DeserializeObject[T](BasicDeliverEventArgs deliverArgs) in /home/vsts/work/1/s/src/common/Smi.Common/MessageSerialization/JsonConvert.cs:line 69
   at Smi.Common.Messaging.Consumer`1.SafeDeserializeToMessage[T](IMessageHeader header, BasicDeliverEventArgs deliverArgs, T& iMessage) in /home/vsts/work/1/s/src/common/Smi.Common/Messaging/Consumer.cs:line 139 0: Required property 'StatusMessage' not found in JSON. Path '', line 1, position 434.|MessageGuid=4d7c4f04-884d-4c5b-ad2c-b5f69605c6d9, ProducerProcessID=1081514896, ProducerExecutableName=CTPAnonymiserHost, OriginalPublishTimestamp=1635439874, Parents=ed762b00-dbba-4239-b645-3d914b7531f6->c6e37e8d-7442-4aef-abc8-6a4119a9b051
2021-10-28 20:22:51.8309|ERROR|ExtractFileCollectionMessageConsumer|Could not deserialize message to ExtractFileCollectionInfoMessage object. Likely an issue with the message content|Newtonsoft.Json.JsonSerializationException: Couldn't deserialize message to Smi.Common.Messages.Extraction.ExtractFileCollectionInfoMessage. See exception data.
   at Smi.Common.MessageSerialization.JsonConvert.DeserializeObject[T](String message) in /home/vsts/work/1/s/src/common/Smi.Common/MessageSerialization/JsonConvert.cs:line 52
   at Smi.Common.MessageSerialization.JsonConvert.DeserializeObject[T](BasicDeliverEventArgs deliverArgs) in /home/vsts/work/1/s/src/common/Smi.Common/MessageSerialization/JsonConvert.cs:line 69
   at Smi.Common.Messaging.Consumer`1.SafeDeserializeToMessage[T](IMessageHeader header, BasicDeliverEventArgs deliverArgs, T& iMessage) in /home/vsts/work/1/s/src/common/Smi.Common/Messaging/Consumer.cs:line 139 0: Required property 'StatusMessage' not found in JSON. Path '', line 1, position 434.|MessageGuid=6999ff9f-5d4a-4dda-9fbd-492b415dd665, ProducerProcessID=30624, ProducerExecutableName=smi, OriginalPublishTimestamp=1635439874, Parents=ed762b00-dbba-4239-b645-3d914b7531f6

```

**Comment by rkm** [comment]
Looks like we may not be setting the StatusMessage properly in all cases [here](https://github.com/SMI/SmiServices/blob/master/src/microservices/com.smi.microservices.ctpanonymiser/src/main/java/org/smi/ctpanonymiser/messaging/CTPAnonymiserConsumer.java#L116-L118). I'm guessing that the call to `_anonTool.anonymize` returns an error status but doesn't set `_lastStatus` to a non-null value. This then gets missed once the message gets serialized to JSON.

There's another issue though. Debugging CohortPackager, and assuming your job ID is `29a5e4a7-8091-44c0-a726-e620dce3587e`, CohortPackager is still `WaitingForCollectionInfo` with 99/100 messages, so it thinks it's waiting for a message from CohortExtractor. The trace logs show it did receive it though:

```conssole
$ grep 'ExtractFileCollectionMessageConsumer|Acknowledged' 2021-10-28-20-22-34-7977-trace.log | wc -l
99
$ grep 'ExtractFileCollectionMessageConsumer|Rec' 2021-10-28-20-22-34-7977-trace.log | wc -l
100
```

And the full error is in the debug log (as you posted above):

```console
2021-10-28 20:22:51.8309|ERROR|ExtractFileCollectionMessageConsumer|Could not deserialize message to ExtractFileCollectionInfoMessage object. Likely an issue with the message content|Newtonsoft.Json.JsonSerializationException: Couldn't deserialize messag
e to Smi.Common.Messages.Extraction.ExtractFileCollectionInfoMessage. See exception data.
   at Smi.Common.MessageSerialization.JsonConvert.DeserializeObject[T](String message) in /home/vsts/work/1/s/src/common/Smi.Common/MessageSerialization/JsonConvert.cs:line 52
   at Smi.Common.MessageSerialization.JsonConvert.DeserializeObject[T](BasicDeliverEventArgs deliverArgs) in /home/vsts/work/1/s/src/common/Smi.Common/MessageSerialization/JsonConvert.cs:line 69
   at Smi.Common.Messaging.Consumer`1.SafeDeserializeToMessage[T](IMessageHeader header, BasicDeliverEventArgs deliverArgs, T& iMessage) in /home/vsts/work/1/s/src/common/Smi.Common/Messaging/Consumer.cs:line 139 0: Required property 'StatusMessage' not 
found in JSON. Path '', line 1, position 434.|MessageGuid=6999ff9f-5d4a-4dda-9fbd-492b415dd665, ProducerProcessID=30624, ProducerExecutableName=smi, OriginalPublishTimestamp=1635439874, Parents=ed762b00-dbba-4239-b645-3d914b7531f6
```

`ExtractFileCollectionInfoMessage` isn't expected to contain a `StatusMessage`, so will have to do more digging there.

**Comment by tznind** [comment]
There is ExtractedFileStatusMessage.StatusMessage

So a service is trying to decode a message as the wrong type?

Or CTP anonymiser microservice is sending to the wrong queue 

Ok the consumer for that is AnonFailedMessageConsumer : Consumer<ExtractedFileStatusMessage>

And also IsIdentifiableQueueConsumer : Consumer<ExtractedFileStatusMessage>, IDisposable

https://github.com/SMI/SmiServices/blob/1160390ffdf99d6548c624072f2de549fbba09c8/src/microservices/Microservices.CohortPackager/Execution/CohortPackagerHost.cs#L105-L110

So we would expect that if anonymisation failed that this consumer would be getting the message about that

![image](https://user-images.githubusercontent.com/31306100/142438805-cff16501-5c4e-4ae2-acb1-4faecea81b3c.png)

**Comment by tznind** [comment]
That would also make sense why you are only seeing 99 acked and stopping because you expect 99 ack on success consumer + 1 ack on the failed message consumer.  Instead we are deserializing the fail on the success channel and never getting to a completion total.

**Comment by rkm** [comment]
Trace log of offending message being received:

```console
$ pwd
/beegfs-hdruk/smi/data/logs/CohortPackager
$ grep 6999ff9f 2021-10-*trace*
2021-10-28-20-22-34-7977-trace.log:2021-10-28 20:22:51.5912|TRACE|ExtractFileCollectionMessageConsumer|Received||MessageGuid=6999ff9f-5d4a-4dda-9fbd-492b415dd665, ProducerProcessID=30624, Pr
oducerExecutableName=smi, OriginalPublishTimestamp=1635439874, Parents=ed762b00-dbba-4239-b645-3d914b7531f6
```

Which shows:
- It's been received by the expected consumer (`ExtractFileCollectionMessageConsumer`)
- `ProducerProcessID=30624` matches an instance of `CohortExtractor`

Log from CohortExtractor shows:

```console
$ pwd
/beegfs-hdruk/smi/data/logs/CohortExtractor
$ grep 6999ff9f 2021-10-28-17-51-42-30624-*
2021-10-28-17-51-42-30624-trace.log:2021-10-28 17:52:00.2312|TRACE|ProducerModel|Sent 6999ff9f-5d4a-4dda-9fbd-492b415dd665 to FileCollectionInfoExchange||MessageGuid=6999ff9f-5d4a-4dda-9fbd-
492b415dd665, ProducerProcessID=30624, ProducerExecutableName=smi, OriginalPublishTimestamp=1635439874, Parents=ed762b00-dbba-4239-b645-3d914b7531f6
```

Which looks as expected, including the exchange it was sent to.

I've also just spotted the same errors in logs from May 18th.

**Comment by tznind** [comment]
The stack trace doesn't look like the deserialization Exception was properly intercepted

https://github.com/SMI/SmiServices/blob/fbb09556bcd297b16a24fd35879b5a8c6dcd3165/src/common/Smi.Common/Messaging/Consumer.cs#L137-L151

**Comment by rkm** [comment]
Fixed with `v5.0.0`
